t/t5799-gvfs-helper.sh

```diff
@@ -0,0 +1,972 @@
+	return 1
+}
+
+start_gvfs_protocol_server_with_mayhem () {
+	if test $# -lt 1
+	then
+		echo "start_gvfs_protocol_server_with_mayhem: need mayhem args"
+		return 1
+	fi
+
+	mayhem=""
+	for k in $*
+	do
+		mayhem="$mayhem --mayhem=$k"
+	done
+	#
+	# Launch our server into the background in repo_src.
+	#
+	(
+		cd "$REPO_SRC"
+		test-gvfs-protocol --verbose \
+			--listen=127.0.0.1 \
+			--port=$GIT_TEST_GVFS_PROTOCOL_PORT \
+			--reuseaddr \
+			--pid-file="$PID_FILE" \
+			$mayhem \
+			2>"$SERVER_LOG" &
+	)
+	#
+	# Give it a few seconds to get started.
+	#
+	for k in 0 1 2 3 4
+	do
+		if test -f "$PID_FILE"
+		then
+			return 0
+		fi
+		sleep 1
+	done
+
+	echo "start_gvfs_protocol_server: timeout waiting for server startup"
+	return 1
+}
+
+# Verify the number of connections from the client.
+#
+# If keep-alive is working, a series of successful sequential requests to the
+# same server should use the same TCP connection, so a simple multi-get would
+# only have one connection.
+#
+# On the other hand, an auto-retry after a network error (mayhem) will have
+# more than one for a single object request.
+#
+# TODO This may generate false alarm when we get to complicated tests, so
+# TODO we might only want to use it for basic tests.
+#
+verify_connection_count () {
+	if test $# -eq 1
+	then
+		expected_nr=$1
+	else
+		expected_nr=1
+	fi
+
+	actual_nr=$(( $(grep "Connection from" "$SERVER_LOG" | wc -l) ))
+
+	if test $actual_nr -ne $expected_nr
+	then
+		echo "verify_keep_live: expected $expected_nr; actual $actual_nr"
+		return 1
+	fi
+	return 0
+}
+
+# Verify that the set of requested objects are present in
+# the shared-cache and that there is no corruption.  We use
+# cat-file to hide whether the object is packed or loose in
+# the test repo.
+#
+# Usage: <pathname_to_file_of_oids>
+#
+verify_objects_in_shared_cache () {
+	#
+	# See if any of the objects are missing from repo_t1.
+	#
+	git -C "$REPO_T1" cat-file --batch-check <"$1" >OUT.bc_actual || return 1
+	grep -q " missing" OUT.bc_actual && return 1
+	#
+	# See if any of the objects have different sizes or types than repo_src.
+	#
+	git -C "$REPO_SRC" cat-file --batch-check <"$1" >OUT.bc_expect || return 1
+	test_cmp OUT.bc_expect OUT.bc_actual || return 1
+	#
+	# See if any of the objects are corrupt in repo_t1.  This fully
+	# reconstructs the objects and verifies the hash and therefore
+	# detects corruption not found by the earlier "batch-check" step.
+	#
+	git -C "$REPO_T1" cat-file --batch <"$1" >OUT.b_actual || return 1
+	#
+	# TODO move the shared-cache directory (and/or the
+	# TODO .git/objects/info/alternates and temporarily unset
+	# TODO gvfs.sharedcache) and repeat the first "batch-check"
+	# TODO and make sure that they are ALL missing.
+	#
+	return 0
+}
+
+verify_received_packfile_count () {
+	if test $# -eq 1
+	then
+		expected_nr=$1
+	else
+		expected_nr=1
+	fi
+
+	actual_nr=$(( $(grep "packfile " OUT.output | wc -l) ))
```

Comment:
How about `actual_nr=$(grep -c "packfile " <OUT.output`?

Reply:
Cool.  A new option I didn't know about.

Reply:
:-) TBH I had not known about it until some time earlier this year, I think. In any case, I still always have to look at https://man.openbsd.org/grep and https://linux.die.net/man/1/grep and sometimes at https://pubs.opengroup.org/onlinepubs/009695399/utilities/grep.html to figure out what is portable, as GNU grep's man pages seem to delight in omitting the information what options are standard compliant and what options are GNU extensions.